buildbot: Python 3 support and other improvements

lopsided98 opened the pull request:
###### Motivation for this change

The latest versions of `pylint` and `astroid` do not support Python 2, which breaks buildbot (#45782) because it only supports Python 2 in nixpkgs. Buildbot supports Python 3, but this is potentially a breaking change because buildbot uses Python in its configuration files. Therefore, this PR adds configurable support for Python 2 or 3, which Python 2 being the default. The 1.0 branches of `pylint` and `astroid` are still supported until Python 2's EOL, so they are used to build buildbot with Python 2.

This PR supersedes #45789.

###### Things done

First, I updated and added Python 3 support to `sphinx-jinja`. This mainly required fixing how the tests were run. This commit supersedes #45763.

I updated `tempita` to use a fork that adds Python 3 support. The upstream tempita package appears to be abandoned. This fork is used by the official Arch Linux package of tempita, but is somewhat undesirable because it is not an official release. This commit supersedes #39692.

I updated `astroid` and `pylint` to their latest versions, and added `astroid1` and `pylint1` packages which use the still maintained 1.0 branch of these packages. Because there are a number of different ways to implement packages with multiple versions in Nix, I'm not sure if I chose the best way.

While I was testing I ran into #40273, which is supposedly fixed but does not seem to be (see https://github.com/NixOS/nixpkgs/issues/40273#issuecomment-417730406). To work around this I changed `hyperlink` to not use pytest to run its tests (the default `checkPhase` seems to work fine) and made some changes to `autobahn` (which do not fix the problem, but only one package needs to be fixed for the conflict to be resolved). These changes are not critical, because they should not be necessary once #40273 is truly fixed, but I included them anyway because I believe they result in cleaner derivations.

I then updated all the components of buildbot to 1.3.0. `buildbot-worker` had already been updated by an automated update, but the rest of the components had not. This has happened a number of times recently and sometimes causes problems, especially when `buildbot-pkg` is updated without updating the plugins.

The last commit adds Python 3 support as well as a number of other improvements that I have been running locally for about a year. There are no intentional breaking changes, but I have not tested every possible scenario.

All the buildbot packages are moved to `pythonPackages`. A new `python` attribute is added to each package's passthru, allowing the NixOS module to determine which version of Python to use. The tests are currently disabled for both Python 2 and 3, because a test fails due to what appears to be a bug in buildbot. I attempted to disable the failing test, but this simply causes the error to move to a different test and so on. The error message (see comment above `doCheck = false`) relates to a recently added buildbot feature, so this points to a problem with buildbot.

I changed how the plugin wrapper works, making the wrapped package into a Python module that can be added to a Python environment using `python.withPackages`.

In the `buildbot-master` NixOS module, I switched to directly invoking `twistd` (as was already done with `buildbot-worker`) in order to log to stdout and therefore the systemd journal, rather than a file. I also added a new `pythonPackages` option, which takes a function that is passed to `python.withPackages` and is used to set the service's `PYTHONPATH`. I also generate the twisted tac file within the module, rather than using `create-master` (`create-master` is still necessary to create the database). This gives more control over the initialization process (especially useful for `buildbot-worker`).

In the `buildbot-worker` NixOS module, I made similar improvements to the `PYTHONPATH` handling, but did not add a `pythonPackages` option because workers never run user Python code AFAIK. I also generate the tac file in the module, which makes it possible to implement the `workerPassFile` option (checks one of the boxes in #24288). I also added `hostMessage` and `adminMessage` options, which allow users to declaratively set those messages.

The `buildbot` NixOS test passes, including the commented out daemon test, which supposedly was previously broken. I configured the tests to run with both Python 2 and 3.

I would appreciate it if other buildbot users would test this PR. I have only run the tests and used it with my buildbot deployment, which has a number of customizations with might obscure bugs.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

lopsided98 committed:
pythonPackages.sphinx-jinja: 0.2.1 -> 1.1.0, Python 3 support

lopsided98 committed:
buildbot: add Python 3 support

lopsided98 committed:
pythonPackages.tempita: 0.5.2 -> 0.5.3-2016-09-28

lopsided98 committed:
python2Packages.astroid: init at 1.6.5

lopsided98 committed:
python2Packages.pylint: init at 1.9.2

lopsided98 committed:
buildbot: 1.2.0 -> 1.4.0

lopsided98 committed:
pythonPackages.hyperlink: don't create .pytest_cache

lopsided98 committed:
pythonPackages.autobahn: use checkInputs

xeji commented:
cc @dotlambda

dotlambda commented on pkgs/development/tools/build-managers/buildbot/default.nix:
```diff
@@ -38,9 +53,10 @@ let
       buildbot-pkg
       treq
+      glibcLocales
     ];
```
Why are the python packages not in `propagatedBuildInputs`?

dotlambda commented on pkgs/development/tools/build-managers/buildbot/default.nix:
```diff
@@ -1,27 +1,42 @@
+
+    passthru = package.passthru // {
+      inherit (package) python;
+      withPlugins = moarPlugins: withPlugins (moarPlugins ++ plugins);
```
`morePlugins`

dotlambda commented on pkgs/development/python-modules/sphinx-jinja/default.nix:
```diff
@@ -0,0 +1,23 @@
+{ lib, buildPythonPackage, fetchPypi, pbr, sphinx-testing, nose, glibcLocales, sphinx }:
+
+buildPythonPackage (rec {
+  name = "${pname}-${version}";
```
Drop this line

dotlambda commented on pkgs/development/python-modules/sphinx-jinja/default.nix:
```diff
@@ -0,0 +1,23 @@
+    # Zip (epub) does not support files with epoch timestamp
+    LC_ALL="en_US.UTF-8" nosetests -e test_build_epub
+  '';
+  meta = with lib; {
```
Please and some blank lines to make this more readable.

dotlambda commented on pkgs/development/python-modules/autobahn/default.nix:
```diff
@@ -12,14 +12,14 @@ buildPythonPackage rec {
+  # Upstream claim python2 support, but tests require pytest-asyncio which
+  # is python3 only. Therefore, tests are skipped for python2.
+  doCheck = isPy3k;
+  checkInputs = [ unittest2 mock pytest pytest-asyncio ];
```
The tests do definitely support python2: https://github.com/crossbario/autobahn-python/blob/master/tox.ini.

FRidh commented on pkgs/development/python-modules/astroid/1.6.nix:
```diff
@@ -0,0 +1,6 @@
+{ callPackage }:
```
If 1.6 is a legacy version we need for Python 2, then we copy in the whole expression, as we don't know how much the new one will deviate from the old one.

FRidh commented on pkgs/development/python-modules/autobahn/default.nix:
```diff
@@ -12,14 +12,14 @@ buildPythonPackage rec {
     (stdenv.lib.optional isPy33 asyncio) ++
     (stdenv.lib.optionals (!isPy3k) [ trollius futures ]);
 
+  # Upstream claim python2 support, but tests require pytest-asyncio which
```
trollius 2 is needed for asyncio support

FRidh commented on pkgs/development/python-modules/pylint/1.9.nix:
```diff
@@ -0,0 +1,7 @@
+{ callPackage, astroid1 }:
```
As with the other package, copy in the whole expression

FRidh commented on pkgs/development/tools/build-managers/buildbot/default.nix:
```diff
@@ -1,27 +1,42 @@
-    ln -sfv ${package}/lib $out/lib
-  '';
+  # Can't use runCommand because it does not support propagatedBuildInputs
+  withPlugins = plugins: toPythonModule (stdenv.mkDerivation {
```
This can simply be `python.withPackages`.

FRidh reviewed (changes requested):
I'm not sure whether we really should move buildbot into python-packages, however, if we do this, then the expressions should live there as well (`python-modules/`), and more importantly, no expressions should use `buildPythonApplication`.

lopsided98 committed:
pythonPackages.sqlalchemy_migrate: patch backslash escapes

lopsided98 committed:
pythonPackages.trollius: 1.0.4 -> 2.2

lopsided98 commented:
I think I have addressed all the issues that were brought up. I had to add a few new commits to fix `autobahn`'s tests with Python 2, and a commit to add a patch to `sqlalchemy_migrate` to make it compatible with Python 3.6 (I previously had this in an overlay and forgot about it).

@nand0p I kept you as the maintainer of the 1.x branches of `astroid` and `pylint`. Is this okay?

Buildbot just updated to 1.4.0, and its tests now pass for Python 2 and 3.

lopsided98 commented:
@FRidh @dotlambda When you get a chance, could you take a look at the changes I made to address your comments?

lopsided98 commented:
I rebased against master and moved the changelog entry to the 19.03 changelog.

Infinisil commented on nixos/modules/services/continuous-integration/buildbot/master.nix:
```diff
@@ -175,18 +199,25 @@ in {
+        default = pythonPackages: with pythonPackages; [ ];
+        defaultText = "pythonPackages: with pythonPackages [ ]";
+        description = "Packages to add the to the PYTHONPATH of the buildbot process.";
+        example = literalExample "pythonPackages: with pythonPackages [ requests ]";
```
Missing `;` after `pythonPackages`

Also in `defaultText`.

Infinisil commented on nixos/modules/services/continuous-integration/buildbot/worker.nix:
```diff
@@ -59,6 +93,24 @@ in {
       };
 
+      workerPassFile = mkOption {
+        default = pkgs.writeText "buildbot-worker-password" cfg.workerPass;
```
This should be moved to the `config` section of this module, because I think this can cause the documentation to be rebuilt when you change the password (because the docs render the default, and the default depends on the config).

So something like
```nix
config = {
  services.buildbot-worker.workerPassFile = mkDefault (pkgs.writeText ...);
}
```

Infinisil commented on pkgs/development/python-modules/tempita/default.nix:
```diff
@@ -0,0 +1,20 @@
+
+  meta = {
+    homepage = https://github.com/gjhiggins/tempita;
+    description = "A very small text templating language";
```
Would be nice to have a license here (ref https://github.com/NixOS/nixpkgs/issues/43716), but the repo doesn't have any and we would have to assume worst-case and make it unfree, but that would be kinda disappointing to have all its dependents be unfree too because of that, so I'd be okay with leaving it out.

lopsided98 replied:
According the original author's archived website (https://web.archive.org/web/20150301124031/http://pythonpaste.org:80/tempita/license.html), it is MIT licensed.

Infinisil commented on pkgs/top-level/all-packages.nix:
```diff
@@ -8128,21 +8128,6 @@ with pkgs;
-  };
-  buildbot-ui = buildbot.withPlugins (with self.buildbot-plugins; [ www ]);
-  buildbot-full = buildbot.withPlugins (with self.buildbot-plugins; [ www console-view waterfall-view grid-view wsgi-dashboards ]);
-
```
Might be worth adding aliases for these in `<nixpkgs/pkgs/top-level/aliases.nix>` for backwards compatibility. But considering it didn't work before, I doubt many people actually use this. Feel free to do whatever you think, I'm fine with or without aliases.

lopsided98 replied:
I added the aliases.

Infinisil commented:
@GrahamcOfBorg test buildbot.python2 buildbot.python3

GrahamcOfBorg commented:
<!--REQUEST_ID=9fb40c0c-0772-4255-8ec1-d738f900e2e4-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45890&attempt_id=cf10592a-4b0d-486a-8ace-322055c7705b)

Attempted: tests.buildbot.python2, tests.buildbot.python3

<details><summary>Partial log (click to expand)</summary><p>

```
Oct 11 22:13:51 bbworker mqa4mvwqafrn53kcwzvxysr4p5llcpmx-unit-script-buildbot-worker-pre-start[753]: mkdir: created directory '/home/bbworker/worker/info'
Oct 11 22:13:51 bbworker systemd[1]: Started Buildbot Worker..
total 12
drwxr-xr-x 3 bbworker bbworker 4096 Oct 11 22:13 .
drwx------ 3 bbworker bbworker 4096 Oct 11 22:13 ..
drwxr-xr-x 2 bbworker bbworker 4096 Oct 11 22:13 info
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/7mnda9bfn2czxnhw4p42hbw4gviwdhb8-vm-test-run-buildbot
/nix/store/qja43p34slaqchzr0pdr5jcd8s90szjn-vm-test-run-buildbot
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=9fb40c0c-0772-4255-8ec1-d738f900e2e4-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45890&attempt_id=40328c57-bcdd-47c1-a6c2-baa612d2842e)

Attempted: tests.buildbot.python2, tests.buildbot.python3

<details><summary>Partial log (click to expand)</summary><p>

```
Oct 11 22:58:14 bbworker systemd[1]: Starting Buildbot Worker....
Oct 11 22:58:14 bbworker systemd[1]: Started Buildbot Worker..
total 12
drwxr-xr-x 3 bbworker bbworker 4096 Oct 11 22:58 .
drwx------ 3 bbworker bbworker 4096 Oct 11 22:58 ..
drwxr-xr-x 2 bbworker bbworker 4096 Oct 11 22:58 info
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/2p09g21258a4dahpyqr9a598gpaif8ij-vm-test-run-buildbot
/nix/store/skbj8hg7in1581fhzm0wll0ycc6w4dd1-vm-test-run-buildbot
```
</p></details>

lopsided98 commented:
@Infinisil I fixed the issues you brought up.

Should this go into staging?

Also, there may be some buildbot users on 18.03 who will find their configs broken when they upgrade to 18.09, so we may want to backport this eventually. I'm not sure if we would backport the whole PR, or just the minimum to fix the build.

Infinisil commented:
I think master is alright. Feel free to open another PR with a backport if you feel it's necessary.

Thanks for the PR, the NixOS test is very nice too :)
